README.md
Spatial frequency preferences
==============================
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/billbrod/spatial-frequency-preferences/v1.0.2?filepath=notebooks)
[![DOI](https://zenodo.org/badge/98347660.svg)](https://zenodo.org/badge/latestdoi/98347660)


An fMRI experiment to determine the relationship between spatial
frequency and eccentricity in the human early visual cortex.

See the [paper](https://doi.org/10.1101/2021.09.27.462032) for scientific
details. If you re-use some component of this project in an academic
publication, see the [citing](#citation) section for how to credit us.

# Usage

The data and code for this project are shared with the primary goal of enabling
reproduction of the results presented in the associated paper. Novel analyses
should be possible, but no guarantees.

To that end, we provide [several entrypoints into the data](#data) for
re-running the full or partial analysis, with a script to automate their
download and proper arrangement. There are also several ways to reproduce the
[computational environment](#software-requirements) of the analysis.

The following steps will walk you through downloading the fully-processed data
and recreating the figures, read further on in this README for details:
1. Clone this repo.
2. Open `config.yml` and modify the `DATA_DIR` path to wherever you wish to
   download the data.
3. Install the python environment:
   - Install [miniconda](https://docs.conda.io/en/latest/miniconda.html) on your
     system for python 3.7.
   - Install [mamba](https://github.com/mamba-org/mamba): `conda install mamba
     -n base -c conda-forge`
   - Navigate to this directory and run `mamba env create -f environment.yml` to
     install the environment.
   - Run `conda activate sfp` to activate the python environment.
4. Run `python download_data.py fully-processed` to download the fully-processed
   data (this is about 500MB).
5. Run `snakemake -k -j N reports/paper_figures/fig-XX.svg`
   (where `N` is the number of cores to use in parallel) to recreate a given
   figure from the paper (note the number must be 0-padded, i.e., `fig-01.svg`,
   *not* `fig-1.svg`). These will end up in the `reports/paper_figures/`
   directory. Note that they are svgs, a vector file format. If your default
   image viewer cannot open them, your browser can. They can be converted to
   pdfs using [inkscape](https://inkscape.org/) or Adobe Illustrator.
   - **WARNING**: while most figures take only a few minutes to create, one of
     these, `fig-08.svg`, takes much longer (up to 8 minutes on the cluster, 21
     minutes on my personal laptop).
6. If you wish to create all the figures from the main body of the text, run
   `snakemake -k -j N main_figure_paper`. If one job fails, this
   will continue to run the others (that's what the `-k` flag means).
   
If you wish to create the supplemental figures as well:
1. Download the additional data required: `python download_data.py supplemental`
   (this is about 5GB). *NOTE*: this is not required if you have already
   downloaded the full `preprocessed` data set from
   [OpenNeuro](https://openneuro.org/datasets/ds003812/).
2. Run `snakemake -k -j N reports/paper_figures/fig-SXX.svg` (where again the
   number must be 0-padded) to create a single supplemental figure or `snakemake
   -k -j N supplement_figure_paper` to create all of them.
   
If you have any trouble with the above, check the
[troubleshooting](#troubleshooting) section to see if there's a solution to your
problem.

To understand what the `snakemake` command is doing, see the [What's going
on?](#whats-going-on) section.
   
## Notebooks

We also include several jupyter notebooks in the `notebooks/` directory:

- `How-to-use-model`: examples of how to use the 2d tuning model described in
  the paper.
- `Stimuli`: some exploration of how to create the stimuli and visualize linear
  approximations thereof.

If you'd like to use them, you can either view it on
[Binder](https://mybinder.org/v2/gh/billbrod/spatial-frequency-preferences/v1.0.2?filepath=notebooks)
or [install
jupyter](https://jupyterlab.readthedocs.io/en/stable/getting_started/installation.html)
so you can view it locally (you can also view a static version of the notebooks
using [nbviewer](https://nbviewer.jupyter.org/)).

If you would like to install jupyter locally and are unfamiliar with it, there
are two main ways of getting it working (after setting up your [conda
environment](#conda-environment)):

1. Install jupyter in this `sfp` environment: 

``` sh
conda activate sfp
conda install -c conda-forge jupyterlab
```

   This is easy, but if you have multiple conda environments and want to use
   Jupyter notebooks in each of them, it will take up a lot of space.
   
2. Use [nb_conda_kernels](https://github.com/Anaconda-Platform/nb_conda_kernels):

``` sh
# activate your 'base' environment, the default one created by miniconda
conda activate 
# install jupyter lab and nb_conda_kernels in your base environment
conda install -c conda-forge jupyterlab
conda install nb_conda_kernels
# install ipykernel in the sfp environment
conda install -n sfp ipykernel
```

   This is a bit more complicated, but means you only have one installation of
   jupyter lab on your machine.
   
In either case, to open the notebooks, navigate to this directory on your
terminal and activate the environment you installed jupyter into (`sfp` for 1,
`base` for 2), then run `jupyter` and open up the notebook. If you followed the
second method, you should be prompted to select your kernel the first time you
open a notebook: select the one named "sfp".

## Model parameters

`data/tuning_2d_model` contains two csv files containing the parameter values
presented in the paper:

- `combined_subject_params.csv` contains the parameter values from combining
  across subjects, as presented in the main body of the paper.
- `individual_subject_params.csv` contains the parameter values from each
  subject individually, as shown in Figure 9B and the appendix.

Both csvs have five columns: fit_model_type, model_parameter, subject, bootstrap_num and fit_value:

- `fit_model_type` (str): the (long) name of this submodel. Both csvs only
  contain a single model, which corresponds to model 9 in our paper, the
  best-fit one.
- `subject` (str): the subject whose parameters these are. In
  combined_subject_params, there's only one subject, 'all'.
- `bootstrap_num` (int): the bootstrap number. Each combination of
  model_parameter and subject has 100 bootstraps. For combined_subject_params,
  these are the bootstraps used to combine across subjects in a
  precision-weighted average. For individual_subject_params, these are the
  bootstraps across fMRI runs. See paper for more details.
- `fit_value` (float): the actual value of the parameter.
- `model_parameter` (str): the (long) name of the model parameter. The following
  gives the mapping between the two:
  - `'sigma'`: <img src="https://latex.codecogs.com/gif.latex?\sigma" />
  - `'sf_ecc_slope'`: <img src="https://latex.codecogs.com/gif.latex?a" />
  - `'sf_ecc_intercept'`: <img src="https://latex.codecogs.com/gif.latex?b" />
  - `'abs_mode_cardinals'`: <img src="https://latex.codecogs.com/gif.latex?p_1" />
  - `'abs_mode_obliques'`: <img src="https://latex.codecogs.com/gif.latex?p_2" />
  - `'rel_mode_cardinals'`: <img src="https://latex.codecogs.com/gif.latex?p_3" />
  - `'rel_mode_obliques'`: <img src="https://latex.codecogs.com/gif.latex?p_4" />
  - `'abs_amplitude_cardinals'`: <img src="https://latex.codecogs.com/gif.latex?A_1" />
  - `'abs_amplitude_obliques'`: <img src="https://latex.codecogs.com/gif.latex?A_2 " />
  - `'rel_amplitude_cardinals'`: <img src="https://latex.codecogs.com/gif.latex?A_3" />
  - `'rel_amplitude_obliques'`: <img src="https://latex.codecogs.com/gif.latex?A_4" />
 
See the [included How-to-use-model notebook](#notebooks) for details about how
to use these csvs with our model.
   
# Setup 

The analyses were all run on Linux (Ubuntu, Red Hat, and CentOS, several
different releases). The steps required to create the figures have also been
tested on Macs, and everything else should work on them as well. For Windows, I
would suggest looking into the [Windows Subsystem for
Linux](https://docs.microsoft.com/en-us/windows/wsl/about), as Windows is very
different from the others, or using the [docker image](#docker-image), which
should work on every OS.

## Software requirements

If you are starting with the partially- or fully-processed data, as explained
[below](#data), then you only need the [python](#python) requirements. If you're
re-running the analysis from the beginning, you will also need [MATLAB](#matlab)
with its associated toolboxes, [FSL, and Freesurfer](#other).

In order to use the included download script `download_data.py` to download
`partially-processed`, `fully-processed`, or `supplemental`, you will also need
`rsync`, which is probably already on your machine.

In order to use `download_data.py` to download the `preprocessed` data from
OpenNeuro, you will need to install the [OpenNeuro command line
interface](https://docs.openneuro.org/packages-openneuro-cli-readme).

### Python

This code works with python 3.6 and 3.7 (it may work with higher versions, but
they haven't been tested).

We provide several ways to reproduce the python environment used in this
experiment, in order of increasing complexity and future-proof-ness:

1. [Conda](#conda-environment)
2. [Docker](#docker-image)
3. Reprozip: under development

#### Conda environment

1. Install [miniconda](https://docs.conda.io/en/latest/miniconda.html) for your
   system with the appropriate python version.
2. Install [mamba](https://github.com/mamba-org/mamba): `conda install mamba
  -n base -c conda-forge`
3. In this directory, run `mamba env create -f environment.yml`.
4. If everything works, type `conda activate sfp` to activate this environment
   and all required packages will be available.
   
As of this writing, few of the required packages have specific version
requirements (those that do are marked in the `environment.yml` file). As this
will probably change in the future, we include a `pip_freeze.txt` file which
shows the installed versions of each python package used to run the analysis
(note this includes more packages than specified in `environment.yml`, because
it includes all of *their* dependencies as well).

If you're re-running the pre-processing, you'll need [WinawerLab's
MRI_tools](https://github.com/WinawerLab/MRI_tools), commit
[8508652bd9e6b5d843d70be0910da413bbee432e](https://github.com/WinawerLab/MRI_tools/tree/8508652bd9e6b5d843d70be0910da413bbee432e).
The results shouldn't change substantially if pre-processed using fMRIPrep.

I recommend using `mamba` instead of `conda` to install the environment because
`conda` tends to hang while attempting to solve the environment.

#### Docker image

If you're having trouble setting up the conda environment, are on Windows, or
are trying to do this far enough in the future that the required package
versions are no longer installable on your machine, you can use the [docker
image](https://hub.docker.com/repository/docker/billbrod/sfp) instead. This only
contains the python requirements (not matlab, FSL, or Freesurfer).

To use this image, we include `docker-compose.yml`, which should hopefully
simplify things:

1. Clone this repo.
2. Make sure you have [docker](https://docs.docker.com/engine/install/) and
   [docker compose](https://docs.docker.com/compose/install/) installed.
3. Change permissions on `.snakemake` and `reports/paper_figures`: `chmod -R 777
   .snakemake reports/paper_figures/` (the docker image will need to write in
   these directories). If you haven't run `snakemake` yet, there will be no

CHANGELOG
# Changes to experiment / stimuli

Several changes were made to the experiment and stimuli through piloting and
development. The data presented in the paper were all `ses-04`, as described
below.

Things that have been constant: stimuli are presented 300 msec on, 200
msec off, with a stream of digits at fixation to use for a 1-back
distractor task. 10 blank "classes" (4 seconds in length) are randomly
interspersed on each run. 12 runs are gathered (or attempted to, first
two sessions ran out of time because of on-the-fly bug-fixing).

1. ses-pilot00 (git commit b88434d6af8cdc92fb741c99954fae05af02f651,
   Aug 23, 2017): negative spiral stimuli had `w_r < 0`, digits for
   the distractor task were shown every stimulus (500 msec). Only
   sub-wlsubj042 scanned using this protocol. Parameters `w_r` and
   `w_a` ran from `2^2.5` to `2^7.5`, with values every half-octave
   (rounded to the nearest integer), for each of the four main
   stimulus classes.
2. ses-pilot01 (git commit 2ab9d11c8ba077f53997dea5e525c53ef9c0dd64,
   Oct 9, 2017): negative spiral stimuli switched to have `w_a < 0`,
   digits are now shown every other stimulus (every
   second). sub-wlsubj001, sub-wlsubj042, and sub-wlsubj045 all
   scanned using this protocol.
3. ses-01, ses-02 (experiment git commit
   3f0920aee0f4f8f198c0f258b63482afbe47e3de, git commit
   aa661f8f0093a7e444fc60796a48d82006679596 for creation of local
   spatial frequency maps for all stimulus types): `w_r` and `w_a` now
   run from `2^2.5` to `2^7` in half-octave steps; this allows us to
   get an extra half-degree closer to the fovea. Also adds constant
   stimuli (spatial frequency constant across the whole image) to
   serve as a sanity check. The two sessions include one with the
   log-polar stimuli, one with the constant (these are referred to ask
   task-sfp and task-sfpconstant, respectively). The stimuli no longer
   have an alpha value and the calculation of their spatial frequency
   is correct (and is correct for the pilot stimuli as well).
4. ses-03 (git commit 69558708537c4d1a82617b05f6e39b4f2c8d7d9a): adds
   16 seconds of blank time at the beginning of each run (to improve
   estimation of baseline) and an extra 8 seconds at the end of each
   run. Still task-sfp (only log-polar stimuli).
5. ses-04 (git commit c7d6ea6543b368f3721ec836e7591f6c86baa438): same
   experimental design as before, but with new task,
   task-sfprescaled. We were concerned about the effect of the scanner
   projector's modulation transfer function, that we might be losing
   contrast at the higher frequencies (because of the projector's
   pointspread function, which effectively acts to blur the
   image). This could result in a lower response to those high spatial
   frequencies, simply from the reduced contrast (rather than the
   higher spatial frequency). In order to test this, we measured the
   projector's MTF (see the
   [spatial-calibration](https://github.com/WinawerLab/spatial-calibration/)
   Github repo for more details) and then constructed our stimuli so
   that their amplitude is rescaled in a spatial frequency-dependent
   manner; the amplitudes of lower spatial frequencies is reduced such
   that, when displayed by the scanner projector, both the high and
   low spatial should have (approximately) the same contrast. We then
   gather these new measurements and run the same analysis to compare
   the effect. We also add the stimuli for task-sfpconstantrescaled
   (same thing for the constant stimuli), but it's unclear if we will
   gather that data (would be ses-05).


sfp/utils.py
#!/usr/bin/python
"""various utils
"""

import os
import numpy as np
import matplotlib.pyplot as plt
import scipy as sp
from . import stimuli as sfp_stimuli
from bids import BIDSLayout
import pandas as pd
from . import first_level_analysis
from . import tuning_curves
import warnings
from . import plotting


def bytescale(data, cmin=None, cmax=None, high=255, low=0):
    """
    Byte scales an array (image).

    Byte scaling means converting the input image to uint8 dtype and scaling
    the range to ``(low, high)`` (default 0-255).

    If the input image already has dtype uint8, no scaling is done.

    This is copied from scipy.misc, where it is deprecated

    Parameters
    ----------
    data : ndarray
        PIL image data array.
    cmin : scalar, optional
        Bias scaling of small values. Default is ``data.min()``.
    cmax : scalar, optional
        Bias scaling of large values. Default is ``data.max()``.
    high : scalar, optional
        Scale max value to `high`.  Default is 255.
    low : scalar, optional
        Scale min value to `low`.  Default is 0.

    Returns
    -------
    img_array : uint8 ndarray
        The byte-scaled array.

    Examples
    --------
    >>> import numpy as np
    >>> from sfp.utils import bytescale
    >>> img = np.array([[ 91.06794177,   3.39058326,  84.4221549 ],
    ...                 [ 73.88003259,  80.91433048,   4.88878881],
    ...                 [ 51.53875334,  34.45808177,  27.5873488 ]])
    >>> bytescale(img)
    array([[255,   0, 236],
           [205, 225,   4],
           [140,  90,  70]], dtype=uint8)
    >>> bytescale(img, high=200, low=100)
    array([[200, 100, 192],
           [180, 188, 102],
           [155, 135, 128]], dtype=uint8)
    >>> bytescale(img, cmin=0, cmax=255)
    array([[91,  3, 84],
           [74, 81,  5],
           [52, 34, 28]], dtype=uint8)
    """
    if data.dtype == np.uint8:
        return data

    if high > 255:
        raise ValueError("`high` should be less than or equal to 255.")
    if low < 0:
        raise ValueError("`low` should be greater than or equal to 0.")
    if high < low:
        raise ValueError("`high` should be greater than or equal to `low`.")

    if cmin is None:
        cmin = data.min()
    if cmax is None:
        cmax = data.max()

    cscale = cmax - cmin
    if cscale < 0:
        raise ValueError("`cmax` should be larger than `cmin`.")
    elif cscale == 0:
        cscale = 1

    scale = float(high - low) / cscale
    bytedata = (data - cmin) * scale + low
    return (bytedata.clip(low, high) + 0.5).astype(np.uint8)


def scatter_heat(x, y, c, **kwargs):
    plt.scatter(x, y, c=c, cmap='RdBu_r', s=50, norm=plotting.MidpointNormalize(midpoint=0),
                vmin=kwargs['vmin'], vmax=kwargs['vmax'])


def create_sin_cpp(size, w_x, w_y, phase=0, origin=None):
    """create a full 2d sine wave, with frequency in cycles / pixel
    """
    if origin is None:
        origin = [(size+1) / 2., (size+1) / 2.]
    x = np.array(range(1, size+1))
    x, y = np.meshgrid(x - origin[0], x - origin[1])
    return np.cos(2*np.pi*x*w_x + 2*np.pi*y*w_y + phase)


def create_sin_cpd(size, w_x_cpd, w_y_cpd, phase=0, stim_rad_deg=12):
    """create a full 2d sine wave, with frequency in cycles / degree

    this converts the desired cycles / degree into the frequency shown in an image by using the
    stim_rad_deg, the radius of the image in degrees of visual angle.
    """
    w_x_pix = w_x_cpd / (size / (2*float(stim_rad_deg)))
    w_y_pix = w_y_cpd / (size / (2*float(stim_rad_deg)))
    return create_sin_cpp(size, w_x_pix, w_y_pix, phase)


def create_circle_mask(x, y, rad, size):
    """create a circular mask

    this returns a circular mask centered at pixel (x, y) with radius rad in a size by size
    image. This can then be multiplied by an image of the same size to mask out everything else.
    """
    x_grid = np.array(range(size))
    x_grid, y_grid = np.meshgrid(x_grid, x_grid)
    mask = np.zeros((size, size))
    mask[(x_grid - x)**2 + (y_grid - y)**2 <= rad**2] = 1
    return mask


def create_ecc_mask(ecc_range, size, max_visual_angle):
    """Create mask selecting range of eccentricities.

    Note this only works for square images.

    Parameters
    ----------
    ecc_range : tuple
        2-tuple of floats specifying the eccentricity of the internal and
        external edges (in that order) of this mask, in degrees.
    size : int
        Height/width of the mask, in pixels.
    max_visual_angle : float
        Height/width of the mask, in degrees.

    Returns
    -------
    mask : np.ndarray
        2d boolean array containing the mask for this eccentricity range.

    """
    if ecc_range[0] >= ecc_range[1]:
        raise Exception("ecc_range[1] must be strictly greater than ecc_range[0]!")
    ppd = float(size) / max_visual_angle
    origin = ((size+1) / 2., (size+1) / 2.)
    min_mask = create_circle_mask(*origin, ecc_range[0]*ppd, size)
    max_mask = create_circle_mask(*origin, ecc_range[1]*ppd, size)
    return np.logical_and(max_mask.astype(bool), ~min_mask.astype(bool))


def create_prf_loc_map(size=1080, max_visual_angle=24, origin=None):
    """Create map of pRF locations

    Default parameters are for the set up used in this experiment

    Parameters
    ----------
    size : int, optional
        Size of the arrays, in pixels. Returned arrays will be 2d, with this
        number of pixels on each dimension.
    max_visual_angle : float, optional
        Diameter of the array, in degrees. Used to convert from pixels to
        degrees.
    origin : tuple or None, optional
        The location of the origin of the image. If None, will place in the
        center.

    Returns
    -------
    eccen, angle : np.ndarray
        The eccentricity and polar angle, at each pixel.

    """
    assert not hasattr(size, '__iter__'), "Only square images permitted, size must be a scalar!"
    size = int(size)
    if origin is None:
        origin = ((size+1) / 2., (size+1) / 2.)
    x, y = np.meshgrid(np.arange(1, size+1) - origin[0],
                       np.arange(1, size+1) - origin[1])
    eccen = np.sqrt(x**2 + y**2)
    angle = np.mod(np.arctan2(y, x), 2*np.pi)
    return eccen, angle


def mask_array_like_grating(masked, array_to_mask, mid_val=128, val_to_set=0):
    """mask array_to_mask the way that masked has been masked

    this takes two square arrays, grating and array_to_mask. masked should already be masked into
    an annulus, while array_to_mask should be unmasked. This then finds the inner and outer radii
    of that annulus and applies the same mask to array_to_mask. the value in the masked part of
    masked should be mid_val (by default, 128, as you'd get when the grating runs from 0 to 255;
    mid_val=0, with the grating going from -1 to 1 is also likely) and the value that you want to
    set array_to_mask to is val_to_set (same reasonable values as mid_val)
    """
    R = sfp_stimuli.mkR(masked.shape) / float(masked.shape[0])
    R_masking = sfp_stimuli.mkR(array_to_mask.shape) / float(array_to_mask.shape[0])
    x, y = np.where(masked != mid_val)
    Rmin = R[x, y].min()
    try:
        array_to_mask[R_masking < Rmin] = val_to_set
    except IndexError:
        # then there is no R<Rmin
        pass
    Rmax = R[x, y].max()
    try:
        array_to_mask[R_masking > Rmax] = val_to_set
    except IndexError:
        # then there is no R>Rmax
        pass
    return array_to_mask


def flat_hyperbola(x, a):
    """hyperbola which is flat until 1 degree
    """
    b = 1.
    period = x*a
    period[x < b] = a*b
    return 1./period


def fit_log_norm(x, y, **kwargs):
    """fit log norm to data and plot the result

    to be used with seaborn.FacetGrid.map_dataframe

    x: string, column in data which contains the x values for this plot.

    y: string, column in data which contains the y values for this plot.

    kwargs must contain `data`, the DataFrame with data to plot.
    """
    data = kwargs.pop('data')
    plot_data = data.groupby(x)[y].mean()

    try:
        popt, pcov = sp.optimize.curve_fit(tuning_curves.log_norm_pdf, plot_data.index,
                                           plot_data.values)
    except RuntimeError:
        # since data is a Series, this is the best way to do this.
        idx = [i for i in data.iloc[0].index if i not in [x, y]]
        warnings.warn("The following was not well fit by a log Gaussian and so is"
                      " skipped:\n%s" % data.iloc[0][idx])
    else:
        plt.plot(plot_data.index, tuning_curves.log_norm_pdf(plot_data.index, *popt), **kwargs)


def fit_log_norm_ci(x, y, ci_vals=[2.5, 97.5], **kwargs):
    """fit log norm to different bootstraps and plot the resulting mean and confidence interval.

    to be used with seaborn.FacetGrid.map_dataframe.

    because this goes through all the bootstraps and calculates their log normal tuning curves
    separately, it's takes much more time than fit_log_norm

    the data passed here must contain a column named `bootstrap_num`, which specifies which number
    bootstrap the observation corresponds to. Each value of bootstrap_num will be fit
    separately. It's recommended (i.e., this function was written assuming), therefore, that your
    data only contains one y value per value of bootstrap_num and value of x.

    x: string, column in data which contains the x values for this plot.

    y: string, column in data which contains the x values for this plot.

    ci_vals: 2-tuple or list of length 2 of floats, optional. the min and max percentile you wish
    to plot as a shaded region. For example, if you wish to plot the 95% confidence interval, then
    ci_vals=[2.5, 97.5] (the default); if you wish to plot the 68%, then ci_vals=[16, 84].

    kwargs must contain `data`, the DataFrame with data to plot.
    """
    data = kwargs.pop('data')
    if 'color' in kwargs:
        color = kwargs.pop('color')
    lines = []
    for boot in data.bootstrap_num.unique():
        plot_data = data.groupby(x)[[y, 'bootstrap_num']]
        plot_data = plot_data.apply(lambda x, j: x[x.bootstrap_num == j], boot)
        plot_idx = plot_data.index.get_level_values(x)
        plot_vals = plot_data[y].values
        try:
            popt, _ = sp.optimize.curve_fit(tuning_curves.log_norm_pdf, plot_idx, plot_vals)
        except RuntimeError:
            # since data is a Series, this is the best way to do this.
            idx = [i for i in data.iloc[0].index if i not in [x, y]]
            warnings.warn("The following bootstrap was not well fit by a log Gaussian and so is"
                          " skipped:\n%s" % data.iloc[0][idx])
        else:
            lines.append(tuning_curves.log_norm_pdf(plot_idx, *popt))
    lines = np.array(lines)
    lines_mean = lines.mean(0)
    cis = np.percentile(lines, ci_vals, 0)
    plt.fill_between(plot_idx, cis[0], cis[1], alpha=.2, facecolor=color, **kwargs)
    plt.plot(plot_idx, lines_mean, color=color, **kwargs)
    return lines


def local_grad_sin(dx, dy, loc_x, loc_y, w_r=None, w_a=None, phase=0, origin=None,
                   stim_type='logpolar'):
    """create a local 2d sin grating based on the gradients dx and dy

    this uses the gradients at location loc_x, loc_y to create a small grating to approximate a
    larger one at that location. This can be done either for the log polar gratings we use as
    stimuli (in which case w_r and w_a should be set to get the phase correct) or a regular 2d sin
    grating (like the one created by create_sin_cpp), in which case they can be left at 0

    dx and dy should be in cycles / pixel

    stim_type: {'logpolar', 'pilot', 'constant'}. what type of stimuli we're creating an
    approximation of (see stimuli.create_sf_maps_cpp for an explanation). used to determine how to
    calculate the local phase.
    """
    size = dx.shape[0]
    x, y = np.meshgrid(np.array(range(1, size+1)) - loc_x,
                       np.array(range(1, size+1)) - loc_y)
    if origin is None:
        origin = ((size+1) / 2., (size+1) / 2.)
    x_orig, y_orig = np.meshgrid(np.array(range(1, size+1))-origin[0],
                                 np.array(range(1, size+1))-origin[1])
    local_x = x_orig[loc_y, loc_x]
    local_y = y_orig[loc_y, loc_x]

    w_x = 2 * np.pi * dx[loc_y, loc_x]
    w_y = 2 * np.pi * dy[loc_y, loc_x]

    # the local phase is just the value of the actual grating at that point (see the explanation in
    # sfp.stimuli._calc_sf_analytically about why this works).
    if stim_type == 'constant':
        if w_r is not None or w_a is not None:
            raise Exception("If stim_type is constant, w_r / w_a must be None!")
        local_phase = np.mod(w_x * local_x + w_y * local_y + phase, 2*np.pi)
    elif stim_type == 'logpolar':
        local_phase = np.mod(((w_r * np.log(2))/2.) * np.log2(local_x**2 + local_y**2) +
                             w_a * np.arctan2(local_y, local_x) + phase, 2*np.pi)
    elif stim_type == 'pilot':
        alpha = 50
        local_phase = np.mod((w_r / np.pi) * np.log2(local_x**2 + local_y**2 + alpha**2) +
                             w_a * np.arctan2(local_y, local_x) + phase, 2*np.pi)
    if w_x == 0 and w_y == 0:
        return 0
    else:
        return np.cos(w_x*x + w_y*y + local_phase)


def find_stim_idx(stim_df, **kwargs):
    stim_df = stim_df.dropna()
    stim_df.class_idx = stim_df.class_idx.astype(int)
    stim_df = stim_df.rename(columns={'index': 'stimulus_index'})
    stim_df = first_level_analysis._add_freq_metainfo(stim_df)
    props = stim_df.copy()
    key_order = ['stimulus_superclass', 'phi', 'freq_space_angle', 'freq_space_distance']
    key_order += [k for k in kwargs.keys() if k not in key_order]
    for k in key_order:
        v = kwargs.get(k, None)
        if v is not None:
            if isinstance(v, str):
                val = v
            else:
                val = props.iloc[np.abs(props[k].values - v).argsort()[0]][k]
            props = props[props[k] == val]
    return props.index[0]


def find_stim_for_first_level(filename, stim_dir):
    if 'pilot00' in filename:
        stim_type = 'pilot'
        stim = np.load(os.path.join(stim_dir, 'task-sfp_ses-pilot00_stimuli.npy'))
        stim_df = pd.read_csv(os.path.join(stim_dir, 'task-sfp_ses-pilot00_stim_description.csv'))
    elif 'pilot01' in filename:
        stim_type = 'pilot'
        stim = np.load(os.path.join(stim_dir, 'task-sfp_ses-pilot01_stimuli.npy'))
        stim_df = pd.read_csv(os.path.join(stim_dir, 'task-sfp_ses-pilot01_stim_description.csv'))
    else:
        if 'task-sfpconstant' in filename:
            stim_type = 'constant'
            stim = np.load(os.path.join(stim_dir, 'task-sfpconstant_stimuli.npy'))
            stim_df = pd.read_csv(os.path.join(stim_dir, 'task-sfpconstant_stim_description.csv'))
        elif 'task-sfprescaled' in filename:
            stim_type = 'logpolar-rescaled'
            stim = np.load(os.path.join(stim_dir, 'task-sfprescaled_stimuli.npy'))
            stim_df = pd.read_csv(os.path.join(stim_dir, 'task-sfprescaled_stim_description.csv'))
        else:
            stim_type = 'logpolar'
            stim = np.load(os.path.join(stim_dir, 'task-sfp_stimuli.npy'))
            stim_df = pd.read_csv(os.path.join(stim_dir, 'task-sfp_stim_description.csv'))
    return {'stim': stim, 'stim_df': stim_df, 'stim_type': stim_type}


def create_data_dict(dataframe_path, stim_dir):
    """given a dataframe path, create the data dictionary

    this data dictionary contains the first level results, the stimuli, the stimuli descriptive
    dataframe, the tuning curves, the stimulus type, and the path to the first level results and
    tuning curve dataframes

    dataframe path can be the path to either the tuning curves or first level analysis
    dataframe. in either case, we'll then attempt to find the other one as well. there's a chance
    that the tuning curves dataframe hasn't been created yet -- in that case, we raise a warning.
    """
    if 'first_level_binned' in dataframe_path:
        first_level_results_path = dataframe_path
        tuning_df_path = first_level_results_path.replace('first_level_binned', 'tuning_curves')
    else:
        tuning_df_path = dataframe_path
        first_level_results_path = tuning_df_path.replace('tuning_curves', 'first_level_binned')
    return_dict = {'df': pd.read_csv(first_level_results_path),
                   'df_filename': first_level_results_path}
    try:
        return_dict.update({'tuning_df': pd.read_csv(tuning_df_path),
                            'tuning_df_filename': tuning_df_path})
    except IOError:
        warnings.warn("Unable to find matching tuning curve dataframe!")
    return_dict.update(find_stim_for_first_level(first_level_results_path, stim_dir))
    return return_dict


def load_data(subject, session=None, task=None, df_mode='full',
              bids_dir='~/Data/spatial_frequency_preferences', **kwargs):
    """load in the first level results, stimuli, stimuli dataframe, and tuning_df; return as a dict

    this loads in first level results and assocaited stimuli and stimuli description dataframe
    for one session.

    you must specify the subject and either the session (in which case we'll load in that one) or
    the task (in which case we ignore all pilot sessions and find the session that matches that
    task).

    df_mode determines whether we load in the full or summary dataframe.

    kwargs should specify other first_level_binned keywords that are specified on the save path:
    vareas (list of ints), eccen (2-tuple of ints, the range of eccentricity considered), eccen_bin
    (boolean), angle_bin (boolean), atlas_type ({"prior", "posterior"}), and mat_type (string
    specifying the type of design matrix).

    bids_dir should be the path to the bids directory

    if more than one file is found satisfying all constraints, then we'll throw an exception.
    """
    bids_dir = os.path.expanduser(bids_dir)
    if 'derivatives' not in bids_dir:
        bids_dir = os.path.join(bids_dir, 'derivatives')
    if 'first_level_binned' not in bids_dir:
        bids_dir = os.path.join(bids_dir, 'first_level_binned')
    stim_dir = os.path.abspath(os.path.join(bids_dir, '..', '..', 'stimuli'))
    subject = subject.replace('sub-', '')
    layout = BIDSLayout(bids_dir, validate=False)
    if session is not None:
        if "ses-" in str(session):
            session = session.replace("ses-", "")
        files = layout.get("file", subject=subject, session=session, suffix=df_mode)
    else:
        if "task-" in task:
            task = task.replace('task-', '')
        # the bit in session is regex to return a string that doesn't start with pilot
        files = layout.get("file", subject=subject, suffix=df_mode, session=r'^((?!pilot).+)')
    for k, v in kwargs.items():
        # we need to treat atlas_type and mat_type, which will show up as a folder in the path,
        # separately from the others, which will just show in the filename of the csv
        if k in ['atlas_type', 'mat_type']:
            val = v
            files = [f for f in files if val in f.split(os.sep)]
        else:
            if k == 'vareas':
                if not isinstance(v, str):
                    val = '-'.join([str(i) for i in v])
                else:
                    val = v
            elif k == 'eccen':
                if not isinstance(v, str):
                    val = '-'.join([str(i) for i in v])
                else:
                    val = v
            elif k == 'eccen_bin':
                if v:
                    val = "eccen_bin"
                else:
                    val = ""
            elif k == 'hemi_bin':
                if v:
                    val = "hemi_bin"
                else:
                    val = ""
            else:
                raise Exception("Don't know how to handle kwargs %s!" % k)
            files = [f for f in files if val in os.path.split(f)[-1]]
    if len(files) != 1:
        raise Exception("Cannot find unique first level results csv that satisfies all "
                        "specifications! Matching files: %s" % files)
    return create_data_dict(files[0], stim_dir)


def _octave_to_degrees(x):
    """Convert octaves to degrees, meant for mapping along 1d tuning curve results."""
    lower = x.tuning_curve_peak * 2**(-x.tuning_curve_bandwidth/2)
    upper = x.tuning_curve_peak * 2**(x.tuning_curve_bandwidth/2)
    return upper - lower
